lib/gpx_ex/parser.ex

```diff
@@ -39,17 +49,57 @@ defmodule GpxEx.Parser do
     }
   end
 
+  defp build_route(route_element) do
+    %GpxEx.Route{
+      name: optional_string(route_element, "name"),
+      points:
+        route_element
+        |> get_route_point_elements()
+        |> Enum.map(&build_waypoint/1)
+    }
+  end
+
+  defp build_waypoint(waypoint_element) do
+    %GpxEx.Waypoint{
+      lat: get_lat(waypoint_element),
+      lon: get_lon(waypoint_element),
+      ele: get_ele(waypoint_element),
+      name: optional_string(waypoint_element, "name"),
+      symbol: optional_string(waypoint_element, "sym"),
+      description: optional_string(waypoint_element, "desc"),
+      url: optional_string(waypoint_element, "url")
+    }
+  end
+
   defp get_track_elements(xml), do: xpath(xml, ~x"//trk"l)
   defp get_segment_elements(xml), do: xpath(xml, ~x"./trkseg"l)
   defp get_point_elements(xml), do: xpath(xml, ~x"./trkpt"l)
 
-  defp get_track_name(xml), do: xpath(xml, ~x"./name/text()"s) |> optinal_string
+  defp get_top_level_waypoint_elements(xml), do: xpath(xml, ~x"//wpt"l)
+
+  defp get_route_elements(xml), do: xpath(xml, ~x"//rte"l)
+  defp get_route_point_elements(xml), do: xpath(xml, ~x"./rtept"l)
 
   defp get_lat(xml), do: xpath(xml, ~x"./@lat"f)
   defp get_lon(xml), do: xpath(xml, ~x"./@lon"f)
-  defp get_ele(xml), do: xpath(xml, ~x"./ele/text()"Fo)
-  defp get_time(xml), do: xpath(xml, ~x"./time/text()"s) |> optinal_string
+  defp get_ele(xml), do: optional_xpath(xml, ~x"./ele/text()"f)
+  defp get_time(xml), do: optional_string(xml, "time")
+
+  defp optional_string(xml, element_type) do
+    el = xpath(xml, ~x"./#{element_type}")
 
-  defp optinal_string(maybe_string) when maybe_string == "",  do: nil
-  defp optinal_string(maybe_string),  do: maybe_string
+    if is_nil(el) do
+      nil
+    else
+      xpath(el, ~x"./text()"s)
+    end
+  end
```

Comment:
I was getting an exception thrown when trying to query `./whatever/text()` when the element in question didn't exist, so I moved to checking for the element, _then_ asking for its text.